### PR TITLE
[RFC] term: Ensure terminal's buffer is still valid in callback functions

### PR DESCRIPTION
**Pull request by jamessan**
When terminal buffers are being quickly created/destroyed, it's possible
for a libvterm callback to be called after the terminal has already been
closed.

`refresh_terminal` is already verifying the `buf_T *` is valid before
using it.  A new `term_buf_valid` function wraps that pattern and should
be used to acquire the `buf_T *` from callbacks.

### CODE REVIEW

File: src/nvim/terminal.c
Comment (justinmk) [inline src/nvim/terminal.c]: could `buf` be NULL here?

File: src/nvim/terminal.c
Comment (justinmk) [inline src/nvim/terminal.c]: `buf->terminal` will still be dangling though. I hope we can prevent `on_job_output` from running instead. It feels like we're patching the symptom here.

Comment (jamessan) [inline src/nvim/terminal.c]: Do you mean `TerminalJobData`'s `term` member?  `on_job_output` doesn't look at the buffer, which may not exist anymore at the point it's called.

Comment (justinmk) [inline src/nvim/terminal.c]: I mean `buf_T.terminal`. I'm thinking of `terminal_destroy()` which won't set `buf->terminal = NULL` if buf_handle is invalid. It seems like we're doing different things in different places, no single source of truth.

I haven't looked closely at all of this yet so I could be off base.

**Comment by nhooyr** [comment]
Seems to fix it for me.

File: src/nvim/terminal.c
Comment (oni-link) [inline src/nvim/terminal.c]: Is the comment still valid? Should `close_buffer()` do something like set `term->buf_handle = 0`?

Comment (jamessan) [inline src/nvim/terminal.c]: > That comment may not be relevant anymore after e897cca,

Actually, it is still relevant in `refresh_terminal_cb` instead of here.

**Comment by jamessan** [comment]
Attempting to make a test for the problem... Current test code is up, but it's not very reliable (even with the fix).  It's ugly -- running a child nvim inside a term so the TUI gets exercised (since that seems to be required to hit the crash).

File: test/functional/terminal/ex_terminal_spec.lua
Comment (justinmk) [inline test/functional/terminal/ex_terminal_spec.lua]: with feed_data(), need to send raw bytes instead of keycodes. e.g. `"\n"` instead of `<CR>`.

File: test/functional/terminal/ex_terminal_spec.lua
Comment (justinmk) [inline test/functional/terminal/ex_terminal_spec.lua]: I don't think wait() will help much here because it only flushes the input from the parent nvim. Doesn't tell us enough about the child nvim.

So this wait() can be removed, or do it only once after the loop.

File: test/functional/terminal/ex_terminal_spec.lua
Comment (justinmk) [inline test/functional/terminal/ex_terminal_spec.lua]: We mostly just want to give ASAN a chance. So maybe 10 iterations instead of 500 is good enough.

All of the screen redraws will cause screen:expect() to be really slow, and cause too many "indetermistic" screen states. Some of the envs on travis are choking. 

We only really care about waiting for the child nvim to process the data, so perhaps try screen:expect() with the "any" parameter:

    screen:expect(tostring(count)..string.rep(' ', 50 - string.len(tostring(count))), 
      nil, nil, nil, true)

File: src/nvim/terminal.c
Comment (justinmk) [inline src/nvim/terminal.c]: perhaps return NULL in the false case, or `buf` in the true case. Then we don't need the `**buf` parameter.

**Comment by justinmk** [comment]
LGTM after comments

File: test/functional/terminal/ex_terminal_spec.lua
Comment (jamessan) [inline test/functional/terminal/ex_terminal_spec.lua]: > We mostly just want to give ASAN a chance. So maybe 10 iterations instead of 500 is good enough.

The conditions where it happens seem to be when there's a backlog of input being processed, so the queue is running faster than the terminal/shell can startup and teardown.  When I manually reproduce it, I stop pressing the x and nvim keeps chugging along for a few more seconds starting/killing terminals before it crashes.

If there's a better way to simulate that stress to increases the chances of hitting the problem, I'm all ears.  If we're not that concerned with exactly hitting it and just want some coverage, then I'm also fine with lowering the counts and simplifying the logic.

Comment (justinmk) [inline test/functional/terminal/ex_terminal_spec.lua]: Either way is fine with me. The main thing then is avoiding the screen:expect() somehow, it's too flaky.

**Comment by jamessan** [comment]
Updated from the review.

File: test/functional/terminal/ex_terminal_spec.lua
Comment (justinmk) [inline test/functional/terminal/ex_terminal_spec.lua]: `-N` is irrelevant for nvim.

File: test/functional/terminal/ex_terminal_spec.lua
Comment (justinmk) [inline test/functional/terminal/ex_terminal_spec.lua]: On some envs this is getting sent to the job stdin before the child nvim has a chance to receive it.

```
[ RUN      ] ...vim/neovim/test/functional/terminal/ex_terminal_spec.lua @ 52: :terminal does not crash when callback fires after terminal closes
./test/functional/ui/screen.lua:298: Failed to match any screen lines.
Expected (anywhere): "50                                                "
Actual:
  |:let g:split = 0                                  |
  |:au TermOpen * nnoremap <buffer> x :q<CR>         |
  |:nnoremap <silent> x :new <bar> let g:split += 1 <|
  |bar> call termopen([&shell])<CR>                  |
  |:set updatetime=0                                 |
  |{1: }                                                 |
  |{3:-- TERMINAL --}                                    |

stack traceback:
	./test/functional/ui/screen.lua:298: in function 'wait'
	./test/functional/ui/screen.lua:209: in function 'expect'
	...vim/neovim/test/functional/terminal/ex_terminal_spec.lua:74: in function <...vim/neovim/test/functional/terminal/ex_terminal_spec.lua:52>
```

In this case the most expedient way to avoid this would be to send the commands as `-c` args.

File: test/functional/terminal/ex_terminal_spec.lua
Comment (jamessan) [inline test/functional/terminal/ex_terminal_spec.lua]: Right, I meant `'-n'`, since I didn't want to type out `'--cmd', 'set noswapfile'`.

**Comment by jamessan** [comment]
Updated to pass the setup commands through `-c` arguments, although I've still seen a failure locally.  I'll look at setting up a `screen:expect`, like `screen_setup` does with the default command.

**Comment by jamessan** [comment]
Ok, so now I'm seeing an occasional `screen:expect` failure on the other end.  There's just a blank screen being matched instead of the expected echo of the count.  I was able to attach to the child nvim and I see that it's trying to exit due to a `SIGHUP`.

<details>
<summary>backtrace</summary>

```
(gdb) bt
#0  0x0000000000a1c9d7 in loop_close (loop=0x2fdfaf0 <main_loop>, wait=true)
    at /home/jamessan/src/github.com/neovim/src/nvim/event/loop.c:87
#1  0x0000000000dec9e9 in event_teardown ()
    at /home/jamessan/src/github.com/neovim/src/nvim/main.c:168
#2  0x000000000123a51d in mch_exit (r=1)
    at /home/jamessan/src/github.com/neovim/src/nvim/os_unix.c:141
#3  0x0000000000e04a6c in getout (exitval=1)
    at /home/jamessan/src/github.com/neovim/src/nvim/main.c:631
#4  0x0000000000f8a9c1 in preserve_exit ()
    at /home/jamessan/src/github.com/neovim/src/nvim/misc1.c:2645
#5  0x00000000012387df in deadly_signal (signum=1)
    at /home/jamessan/src/github.com/neovim/src/nvim/os/signal.c:133
#6  0x00000000012384f6 in on_signal (handle=0x2fd77e0 <shup>, signum=1, data=0x0)
    at /home/jamessan/src/github.com/neovim/src/nvim/os/signal.c:158
#7  0x0000000000a38f9c in signal_event (argv=0x7ffcaa5d67d0)
    at /home/jamessan/src/github.com/neovim/src/nvim/event/signal.c:44
#8  0x0000000000a22770 in multiqueue_process_events (this=0x60400000da90)
    at /home/jamessan/src/github.com/neovim/src/nvim/event/multiqueue.c:146
#9  0x0000000000a1c467 in loop_poll_events (loop=0x2fdfaf0 <main_loop>, ms=0)
    at /home/jamessan/src/github.com/neovim/src/nvim/event/loop.c:56
#10 0x00000000012245f4 in os_breakcheck ()
    at /home/jamessan/src/github.com/neovim/src/nvim/os/input.c:152
#11 0x0000000000d2af39 in vgetorpeek (advance=1)
    at /home/jamessan/src/github.com/neovim/src/nvim/getchar.c:1695
#12 0x0000000000d284b4 in vgetc () at /home/jamessan/src/github.com/neovim/src/nvim/getchar.c:1396
#13 0x0000000000d39dfd in safe_vgetc ()
    at /home/jamessan/src/github.com/neovim/src/nvim/getchar.c:1516
#14 0x00000000016fc463 in state_enter (s=0x7ffcaa5d7fc0)
    at /home/jamessan/src/github.com/neovim/src/nvim/state.c:37
#15 0x0000000001029d8c in normal_enter (cmdwin=false, noexmode=false)
    at /home/jamessan/src/github.com/neovim/src/nvim/normal.c:463
#16 0x0000000000defd35 in main (argc=14, argv=0x7ffcaa5d88b8)
    at /home/jamessan/src/github.com/neovim/src/nvim/main.c:540
(gdb) info threads
  Id   Target Id         Frame
* 1    Thread 0x7f22ccc70bc0 (LWP 11994) "nvim" 0x0000000000a1c9d7 in loop_close (
    loop=0x2fdfaf0 <main_loop>, wait=true)
    at /home/jamessan/src/github.com/neovim/src/nvim/event/loop.c:87
```

</summary>

**Comment by justinmk** [comment]
I wonder if #5986 would fix that.

**Comment by jamessan** [comment]
Doesn't appear to.  Another interesting data point -- we're in a tight loop there.  nvim is consuming all CPU when this happens.

**Comment by justinmk** [comment]
Your test is just _too good_ :) Perhaps `updatetime=0` is too much to handle. Maybe `updatetime=1` helps?

**Comment by jamessan** [comment]
I've confirmed that the test can reproduce the ASAN failure without explicitly setting `'updatetime'`, so I'll remove that.  Unfortunately, that doesn't fix the test failure.

**Comment by jamessan** [comment]
Looking at this again, we're stuck here:

```
(gdb) frame
#0  loop_close (loop=0x2ff3cb0 <main_loop>, wait=true) at /home/jamessan/src/github.com/neovim/src/nvim/event/loop.c:88
(gdb) l
83        uv_close((uv_handle_t *)&loop->children_kill_timer, NULL);
84        uv_close((uv_handle_t *)&loop->poll_timer, NULL);
85        uv_close((uv_handle_t *)&loop->async, NULL);
86        do {
87          uv_run(&loop->uv, wait ? UV_RUN_DEFAULT : UV_RUN_NOWAIT);
88        } while (uv_loop_close(&loop->uv) && wait);
89        multiqueue_free(loop->fast_events);
90        multiqueue_free(loop->thread_events);
91        multiqueue_free(loop->events);
92        kl_destroy(WatcherPtr, loop->children);
```

`uv_loop_close(&loop->uv)` always returns `UV_EBUSY` and wait is true.

This looks a bit like libuv/libuv#765, which suggests that there's something we haven't properly stopped.  Note, there is a zombie child process

```
12146 pts/4    t+    21:37  |                               \_ /home/jamessan/src/github.com/neovim/build/bin/nvim -u NONE -i NONE -N --cmd set shortmess+=I background=light noswapfile noautoindent l
12165 ?        Zs     0:00  |                               |   \_ [zsh] <defunct>
```

from the shell being spawned by the terminal.

Here's a dump of `loop->uv`, in case that helps:

```
$13 = {
  data = 0x2ff3cb0 <main_loop>,
  active_handles = 0,
  handle_queue = {0x2ff3f38 <main_loop+648>, 0x2febf00 <shup+32>},
  active_reqs = {0x2ff3cd0 <main_loop+32>, 0x2ff3cd0 <main_loop+32>},
  stop_flag = 0,
  flags = 0,
  backend_fd = 10,
  pending_queue = {0x2ff3cf8 <main_loop+72>, 0x2ff3cf8 <main_loop+72>},
  watcher_queue = {0x2ff3d08 <main_loop+88>, 0x2ff3d08 <main_loop+88>},
  watchers = 0x611000009a00,
  nwatchers = 30,
  nfds = 2,
  wq = {0x2ff3d28 <main_loop+120>, 0x2ff3d28 <main_loop+120>},
  wq_mutex = {
    __data = {
      __lock = 0,
      __count = 0,
      __owner = 0,
      __nusers = 0,
      __kind = 0,
      __spins = 0,
      __elision = 0,
      __list = {
        __prev = 0x0,
        __next = 0x0
      }
    },
    __size = '\000' <repeats 39 times>,
    __align = 0
  },
  wq_async = {
    data = 0x0,
    loop = 0x2ff3cb0 <main_loop>,
    type = UV_ASYNC,
    close_cb = 0x0,
    handle_queue = {0x2febf00 <shup+32>, 0x2ff3f38 <main_loop+648>},
    u = {
      fd = 0,
      reserved = {0x0, 0x0, 0x0, 0x0}
    },
    next_closing = 0x0,
    flags = 49152,
    async_cb = 0x7f7f5d0e9f80,
    queue = {0x2ff3e60 <main_loop+432>, 0x2ff3e60 <main_loop+432>},
    pending = 0
  },
  cloexec_lock = {
    __data = {
      __lock = 0,
      __nr_readers = 0,
      __readers_wakeup = 0,
      __writer_wakeup = 0,
      __nr_readers_queued = 0,
      __nr_writers_queued = 0,
      __writer = 0,
      __shared = 0,
      __rwelision = 0 '\000',
      __pad1 = "\000\000\000\000\000\000",
      __pad2 = 0,
      __flags = 0
    },
    __size = '\000' <repeats 55 times>,
    __align = 0
  },
  closing_handles = 0x0,
  process_handles = {0x2ff3e20 <main_loop+368>, 0x2ff3e20 <main_loop+368>},
  prepare_handles = {0x2ff3e30 <main_loop+384>, 0x2ff3e30 <main_loop+384>},
  check_handles = {0x2ff3e40 <main_loop+400>, 0x2ff3e40 <main_loop+400>},
  idle_handles = {0x2ff3e50 <main_loop+416>, 0x2ff3e50 <main_loop+416>},
  async_handles = {0x2ff3dc8 <main_loop+280>, 0x2ff3dc8 <main_loop+280>},
  async_watcher = {
    cb = 0x7f7f5d0ebed0,
    io_watcher = {
      cb = 0x7f7f5d0ebfb0,
      pending_queue = {0x2ff3e80 <main_loop+464>, 0x2ff3e80 <main_loop+464>},
      watcher_queue = {0x2ff3e90 <main_loop+480>, 0x2ff3e90 <main_loop+480>},
      pevents = 1,
      events = 1,
      fd = 13
    },
    wfd = -1
  },
  timer_heap = {
    min = 0x0,
    nelts = 0
  },
  timer_counter = 38,
  time = 4646119503,
  signal_pipefd = {11, 12},
  signal_io_watcher = {
    cb = 0x7f7f5d0f3de0,
    pending_queue = {0x2ff3ee8 <main_loop+568>, 0x2ff3ee8 <main_loop+568>},
    watcher_queue = {0x2ff3ef8 <main_loop+584>, 0x2ff3ef8 <main_loop+584>},
    pevents = 1,
    events = 1,
    fd = 11
  },
  child_watcher = {
    data = 0x0,
    loop = 0x2ff3cb0 <main_loop>,
    type = UV_SIGNAL,
    close_cb = 0x0,
    handle_queue = {0x2ff3d80 <main_loop+208>, 0x2ff3cc0 <main_loop+16>},
    u = {
      fd = 0,
      reserved = {0x0, 0x0, 0x0, 0x0}
    },
    next_closing = 0x0,
    flags = 32768,
    signal_cb = 0x0,
    signum = 0,
    tree_entry = {
      rbe_left = 0x0,
      rbe_right = 0x0,
      rbe_parent = 0x0,
      rbe_color = 0
    },
    caught_signals = 0,
    dispatched_signals = 0
  },
  emfile_fd = 14,
  inotify_read_watcher = {
    cb = 0x0,
    pending_queue = {0x0, 0x0},
    watcher_queue = {0x0, 0x0},
    pevents = 0,
    events = 0,
    fd = 0
  },
  inotify_watchers = 0x0,
  inotify_fd = -1
}
```

**Comment by justinmk** [comment]
I observed the same thing when testing `yes` in a `:term`, for the ctrl-c test. Which is similar to a tight loop. The ctrl-c test actually crashes nvim sometimes.

**Comment by jamessan** [comment]
I think some of the recent term-related fixes may have fixed the test problems I was seeing.

**Comment by justinmk** [comment]
The `tui with non-tty file descriptors can handle pipes as stdout and stderr` failure may be my fault, just ignore it for now. But the other Quickbuild failure in that same log is the test added here.

**Comment by jamessan** [comment]
Since #6644 has made the [failure] I was seeing much more prevalent, I'm inclined to merge this.  I'll re-push later today to get another QB run, but pending any new issues or objections I'll merge this in the next day or so.

[failure]: https://github.com/neovim/neovim/pull/6195#issuecomment-289188508

**Review by justinmk — DISMISSED** [review]

File: src/nvim/terminal.c
Comment (bfredl) [inline src/nvim/terminal.c]: can't we make sure buffer is deleted from the handle dict at the same time as it is invalidated? Otherwise I would expect similar bug to sneak up sooner or later...

Comment (jamessan) [reply src/nvim/terminal.c]: Something like this?

```diff
diff --git a/src/nvim/buffer.c b/src/nvim/buffer.c
index b5ca6543c..b32c33869 100644
--- a/src/nvim/buffer.c
+++ b/src/nvim/buffer.c
@@ -459,6 +459,7 @@ void close_buffer(win_T *win, buf_T *buf, int action, int abort_if_last)
 
   if (buf->terminal) {
     terminal_close(buf->terminal, NULL);
+    handle_unregister_buffer(buf);
   }
 
   /* Always remove the buffer when there is no file name. */
```

`handle_unregister_buffer()` is already called by `close_buffer()` (via `free_buffer()`), but it happens a bit later in the function.  My only question here would be whether this introduces other issues because there are other parts of `close_buffer()` that expect the handle to still be registered.

Based on current test coverage, that doesn't seem to be the case, but that doesn't mean there isn't or that there won't be something added in the future which does need it.

Comment (bfredl) [reply src/nvim/terminal.c]: Adding more _ad-hoc_ logic like that certainly doesn't address my concern in any way, if anything it will make the code _harder_ to reason about... My point was that _invalid_ buffer handles have nothing to do in the buffer table, under _any condition_. Buffer handles are something introduces by nvim, I wouldn't expect vim's destruction code to rely on them in any way. 

>  that there won't be something added in the future which does need it.

And why would we add code that relies on invalid handles after we just removed them? Exactly what would be the point of code that relies on invalid memory?

Comment (jamessan) [reply src/nvim/terminal.c]: Well, it isn't invalid memory until it gets freed.  There are autocmds that run during `close_buffer`.  If any of those end up calling code that requires the handle, then they won't work anymore.

Comment (bfredl) [reply src/nvim/terminal.c]: I see, so it is necessary from a backwards compatibility point-of-view that arbitarary code will get exposed to a _partially invalidated_ buffer object...

**Comment by justinmk** [comment]
> The `tui with non-tty file descriptors can handle pipes as stdout and stderr` failure

Noticed that failure in #7720 where `Process exited 134` (SIGABRT) appeared in the buffer. Maybe obvious, but I guess what happens is that `execvp()` fails (after which we `abort()`). Then the buffer acts like a "normal" buffer: the input that was intended for the job is inserted as text.

**Comment by justinmk** [comment]
> `uv_loop_close(&loop->uv)` always returns `UV_EBUSY` and wait is true.

@jamessan the hang should be fixed now. #11887

**Comment by justinmk** [comment]
@jamessan my comment was meant to suggest that this could be worth rebasing, since before https://github.com/neovim/neovim/pull/11887 this PR was blocked on the hang  :) Or do you think this isn't relevant anymore?

**Pull request closed without merging**
